cdap-ui/app/cdap/components/AbstractWidget/AbstractMultiRowWidget/AbstractRow.tsx

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+import * as React from 'react';
+import IconButton from '@material-ui/core/IconButton';
+import AddIcon from '@material-ui/icons/Add';
+import DeleteIcon from '@material-ui/icons/Delete';
+import { WithStyles } from '@material-ui/core/styles/withStyles';
+import If from 'components/If';
+
+enum KEY_CODE {
+  Enter = 13,
+  Up = 38,
+  Down = 40,
+}
+
+const styles = (theme) => {
+  return {
+    root: {
+      height: '44px',
+    },
+  };
+};
+
+export interface IAbstractRowProps<S extends typeof styles> extends WithStyles<S> {
+  value: string;
+  id: string;
+  index: number;
+  autofocus: boolean;
+  disabled: boolean;
+  onChange: (id: string, value: string) => void;
+  addRow: () => void;
+  removeRow: () => void;
+  changeFocus: (index: number) => void;
+}
+
+export interface IAbstractRowState {
+  value?: string;
+}
+
+export default class AbstractRow<
+  P extends IAbstractRowProps<typeof styles>,
+  State extends IAbstractRowState
+> extends React.PureComponent<P, State> {
+  public static defaultProps = {
+    valuePlaceholder: 'Value',
+  };
+
+  public state = {
+    value: this.props.value,
+  } as State;
+
+  public onChange = (e) => {
+    const value = e.target.value;
```

Comment:
One suggestion would be to have the value being passed into this function. This way if any subsequent child class can process the value and just call onChange. 
Right now we are calling `this.props.onChange(this.props.id, value)` in both `AbstractRow` and in `KeyValueRow` components. Instead if the `handleChange` function just pre-processed the value and called the `onChange` in `AbstractRow` it will be more predictable.

In the future if we have `KeyMultipleValues` or `KeyEncodedValues` widgets they can still do the same `handleChange` which pre-processes the user input and call `onChange` here to propagate the value up.